### ISSUE TITLE
fix: memory leak due to design token bindings

### ISSUE THREAD

**Issue by m-akinc** [microsoft/fast#7022]
# 🐛 Bug Report

Given a component whose template uses a derived token value, as instances of that component are replaced in the DOM, the old, detached instances cannot be garbage collected, i.e. are **leaked**.

## 💻 Repro or Code Sample

See [this codepen](https://codepen.io/m-akinc/pen/JjQzbZj). Each time you click the button, it gets replaced by a new button instance. The old buttons, which are no longer referenced from the code or DOM, are permanently leaked. See below:

![fast-memory-leak](https://github.com/user-attachments/assets/9586c19f-1ba8-4411-b17a-a4a6eb0bdc88)

## 🤔 Expected Behavior

Detached components should be garbage collected.

## 😯 Current Behavior

Memory is leaked.

## 💁 Possible Solution

The cause seems to be that `DesignTokenNode`s do not tear down their binding observers when their corresponding nodes are detached. I.e. this problem is resolved by adding the below `for` loop to `DesignTokenNode.unbind()`:
```ts
    /**
     * Invoked when the DesignTokenNode.target is detached from the document
     */
    public unbind(): void {
        if (this.parent) {
            const parent = childToParent.get(this)!;
            parent.removeChild(this);
        }
        for (const token of this.bindingObservers.keys()) {
            this.tearDownBindingObserver(token);
        }
    }
```

## 🔦 Context

We have a custom table component with row components that are replaced when scrolled in and out of view. This quickly eats up large amounts of memory.

**Comment by chrisdholt** [comment]
Thanks @m-akinc - given the nature of this (memory leak), we'd welcome a PR to FE1 to help resolve this for you.

**Comment by m-akinc** [comment]
> Thanks @m-akinc - given the nature of this (memory leak), we'd welcome a PR to FE1 to help resolve this for you.

Here's that PR: #7023

**Comment by rajsite** [comment]
@chrisdholt we have validated the memory leak issues are resolved with #7023 included in `@microsoft/fast-foundation@2.50.0`.

We did find changes to performance in browsers as a result of the changes but with [some investigation](https://github.com/ni/nimble/issues/2456) we are pretty confident the changes are representative of the GC being able to do more collections enabled by the leak clean-up (observed changes in performance seem to be related to GC pauses and not changes in script runtime).

We have not (yet?) hit the [stack overflow regression discussed on discord](https://discord.com/channels/449251231706251264/449251231706251266/1303079745738510336) but as there are multiple changes included in the `@microsoft/fast-foundation@2.50.0` release additional investigation would be needed to determine if there are other bugfixes warranted.

I think this issue can can be closed and further investigations would be separate issues.

**Comment by chrisdholt** [comment]
> [@chrisdholt](https://github.com/chrisdholt) we have validated the memory leak issues are resolved with [#7023](https://github.com/microsoft/fast/pull/7023) included in `@microsoft/fast-foundation@2.50.0`.
> 
> We did find changes to performance in browsers as a result of the changes but with [some investigation](https://github.com/ni/nimble/issues/2456) we are pretty confident the changes are representative of the GC being able to do more collections enabled by the leak clean-up (observed changes in performance seem to be related to GC pauses and not changes in script runtime).
> 
> We have not (yet?) hit the [stack overflow regression discussed on discord](https://discord.com/channels/449251231706251264/449251231706251266/1303079745738510336) but as there are multiple changes included in the `@microsoft/fast-foundation@2.50.0` release additional investigation would be needed to determine if there are other bugfixes warranted.
> 
> I think this issue can can be closed and further investigations would be separate issues.

Thanks for following up here!
